Update muspinsim tools to v2.0.2 #83

joelvdavies opened the pull request:
Updates the generate and simulate tool wrappers for muspinsim to v2.0.2.

patrick-austin reviewed (commented):
Really like the move to generalise the test data files.

On the surface of it, it all looks OK, but the `average_axes_options_alc` got me thinking:

- Why do we need a separate set of options here? I assume because we cannot set an intrinsic field for the ALC example so having the option to average it is meaningless?
- If so, why are we able to average over polarisation if alc also sets that to one value and doesn't let us edit it? Similarly zero field restricts options for field, intrinsic field and polarisation, but that has all options available to average?

I tried to experiment with it but unfortunately kept getting the error: `trace not found` - possibly due to my dev setup but I haven't had time to look into it yet (or look into what MSS does to try and work out how the averages would behave in these circumstances).

Maybe (i.e. if there's not a simple answer to all this) it would be worth having a chat about this on Monday @joelvdavies ?

joelvdavies commented:
> Really like the move to generalise the test data files.
> 
> On the surface of it, it all looks OK, but the `average_axes_options_alc` got me thinking:
> 
> * Why do we need a separate set of options here? I assume because we cannot set an intrinsic field for the ALC example so having the option to average it is meaningless?
> * If so, why are we able to average over polarisation if alc also sets that to one value and doesn't let us edit it? Similarly zero field restricts options for field, intrinsic field and polarisation, but that has all options available to average?
> 
> I tried to experiment with it but unfortunately kept getting the error: `trace not found` - possibly due to my dev setup but I haven't had time to look into it yet (or look into what MSS does to try and work out how the averages would behave in these circumstances).
> 
> Maybe (i.e. if there's not a simple answer to all this) it would be worth having a chat about this on Monday @joelvdavies ?

Yes, so I basically just added the extra options but left the rest as is. The ALC experiment type sets a range of values (https://muon-spectroscopy-computational-project.github.io/muspinsim/input/#experiment) and the `intrinsic_field` is ignored in this case. As for zero field and the options we can average over but not set, these options have no effect so I will remove them.
Not sure about that error though, mine works locally, do you know where it came from?

bgruening commented on muspinsim_config/muspinsim_config.xml:
```diff
@@ -69,248 +71,166 @@
-                            help="Define 1X3 vector for local magnetic field coupling (T).
-                            Allows default expressions, constants and functions (see help)"/>
+                        <param name="zeeman_index" type="integer" value="" label="Index of coupled spin" min="1" help="Index refers to the order of the spins listed in 'spins' section.  The first spin in the list has index 1, the second has index 2, and so on."/>
+                        <param name="zeeman_vector" type="text" value="" label="Zeeman coupling vector" help="Define 1X3 vector for local magnetic field coupling (T). Allows default expressions, constants and functions (see help)"/>
```
Maybe here should a validator be added? I think there was already one in some other tools that checked a matrix like input.

joelvdavies replied:
Thanks for the suggestion. I think it’s more complicated than its worth here since this input should not only take in digits separated by spaces but also variables and basic python expressions.

patrick-austin reviewed (changes requested):
I've tried out the experimental parameter options and run into a number of silent issues with allowing the user to choose values that actually don't do anything, because we overwrite them with something else in the config. Most of these are hit my individual comments, but there are two other general points with less obvious solutions (let me know what your feelings on them are):

1. For custom experiments, we will either use `time` or `field` as the x axis, meaning we cannot average over that variable. However both of these will appear in the list of `average_axes` regardless. In an ideal world I think we would either not display the option that is currently selected for the x axis, or (perhaps easier to implement) use a `<validator>` to error when the same option is selected as x axis and average. 
2. The ordering of experimental section could be more consistent. Currently the order of the averages, celio, and experimental variables (field temp etc.) varies between each of the presets. I think a consistent approach would give a better UX. For example, all the variables appearing one after the other uninterupted (and in the same order), then the option to select the axes to average, then the option to use celio last when relevant. I appreciate this will make it harder to re-use code, but I think migrating the parameter xml to the macros file (like we now have for `times`) will allow us to then order things in the main xml without too much pain.

It's also worth saying that MuSpinSim is where a lot of this behaviour comes from (e.g. it will overwrite the x axis for ALC even if I'm running from the command line, not just in Galaxy), BUT I think that we have a chance to (and should) leverage the Galaxy UI to make the configuration easier (by making removing config options that would get overridden, explicitly stating what the presets will do etc.)

joelvdavies commented on muspinsim_config/build_file.py:
```diff
@@ -297,14 +313,10 @@ def parse_spin(spin):
-    "experiment_preset": lambda value: build_block(
-        "experiment", [value]
-    ),
+    "average_axes": lambda values: build_block("average_axes", values),
```
Just checked, yes none will do it, its case insensitive.

joelvdavies commented:
> I've tried out the experimental parameter options and run into a number of silent issues with allowing the user to choose values that actually don't do anything, because we overwrite them with something else in the config. Most of these are hit my individual comments, but there are two other general points with less obvious solutions (let me know what your feelings on them are):
> 
> 1. For custom experiments, we will either use `time` or `field` as the x axis, meaning we cannot average over that variable. However both of these will appear in the list of `average_axes` regardless. In an ideal world I think we would either not display the option that is currently selected for the x axis, or (perhaps easier to implement) use a `<validator>` to error when the same option is selected as x axis and average.
> 2. The ordering of experimental section could be more consistent. Currently the order of the averages, celio, and experimental variables (field temp etc.) varies between each of the presets. I think a consistent approach would give a better UX. For example, all the variables appearing one after the other uninterupted (and in the same order), then the option to select the axes to average, then the option to use celio last when relevant. I appreciate this will make it harder to re-use code, but I think migrating the parameter xml to the macros file (like we now have for `times`) will allow us to then order things in the main xml without too much pain.
> 
> It's also worth saying that MuSpinSim is where a lot of this behaviour comes from (e.g. it will overwrite the x axis for ALC even if I'm running from the command line, not just in Galaxy), BUT I think that we have a chance to (and should) leverage the Galaxy UI to make the configuration easier (by making removing config options that would get overridden, explicitly stating what the presets will do etc.)

I have adjusted the layout so it should be a more consistent order between experiment types. For 1 I have modified the options to remove the invalid options from average axes, and similarly removed the time option for the x axis when integral is selected for the y axis.

patrick-austin reviewed (changes requested):
Great work fixing that issue by adding `average_axes_restricted` to `build_file.py`, I think we just need something similar here to account for the different y axis cases.

joelvdavies commented on muspinsim_config/build_file.py:
```diff
@@ -295,16 +311,19 @@ def parse_spin(spin):
+        "x_axis_options": {
+            "x_axis": lambda value: build_block("x_axis", [value]),
+            "average_axes": lambda values: build_block("average_axes", values),
+        },
```
Thanks for spotting that

patrick-austin commented on muspinsim_config/muspinsim_config.xml:
```diff
@@ -375,42 +234,80 @@
+            <output name="out_file" file="test_6.in" ftype="txt" compare="diff"/>
+        </test>
+        <test expect_num_outputs="1">
+            <param name="out_file_prefix" value="test_6"/>
```
```suggestion
            <param name="out_file_prefix" value="test_7"/>
```

patrick-austin commented on muspinsim_config/muspinsim_config.xml:
```diff
@@ -375,42 +234,80 @@
+        </test>
+        <test expect_num_outputs="1">
+            <param name="out_file_prefix" value="test_6"/>
+            <param name="experiment_preset" value="zero_field"/>
```
```suggestion
            <param name="experiment_preset" value="custom"/>
```
Can we make this a test of `custom`? The bug I observed was only with that setup, currently this test wouldn't reproduce the bug since it's using the `zero_field` logic like 6.

patrick-austin commented on muspinsim_config/test-data/test_7.in:
```diff
@@ -0,0 +1,19 @@
+#######################################################
+
+name
+    test_6
```
```suggestion
    test_7
```

patrick-austin commented on muspinsim_config/test-data/test_7.in:
```diff
@@ -0,0 +1,19 @@
+    orientation
+    temperature
+experiment
+    zero_field
```
```suggestion
    custom
```
Would also hope to see sections for `x_axis` and `y_axis` once we're testing `custom`, but not sure of the order to expect. Probably easiest to just regenerate the file to match the test output as long is it's as expected.

patrick-austin reviewed (changes requested):
I've checked manually and can't find any issue with the axes for any experimental option - just want to make sure the new test (7) covers the case that originally led to the bug to avoid regression.

joelvdavies commented:
> Appreciate this was only supposed to be updating the wrappers (sorry for getting it bogged down in the axes stuff), but I think the UX will be better as a result of the changes you've made. The config options are clearer and more interactive, which I think is a real advantage of using the Galaxy UI over CL and writing config by hand.

No worries, I totally agree, it was a good idea.
